### PR TITLE
Feature/farms portfolio

### PR DESCRIPTION
**Pull request by memacius32**
Adding Farms portfolio:
 - Total currently deposited in farms.
 - Total rewards by token.

### CODE REVIEW

File: src/components/earn/FarmBanner/index.tsx [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+      lpToken: farm.lpToken,
+      userLPStakedAmount: farm.userLPStakedAmount,
+      totalPoolAmountUSD: farm.totalPoolAmountUSD,
+      chefVersion: farm.chefVersion
+    }) ?? {}
+
+  const stakingInfo = useSingleFarm(Number(farmId))
+  const { chefVersion, doubleRewardAmount, earnedAmount, doubleRewardToken } = stakingInfo
+
+  const isDualRewards = chefVersion === 1
```
Comment (specialist0x) [inline src/components/earn/FarmBanner/index.tsx:45]: This is `==` on the other file -- is the response always a number?

File: src/components/earn/FarmBanner/index.tsx [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+  const newDoubleRewardsAmount =
+    isDualRewards && isTokenAmountPositive(doubleRewardAmount) ? doubleRewardAmount.toFixed(6) : 0
+  const parsedNewDoubleRewardsAmount = newDoubleRewardsAmount ? parseFloat(newDoubleRewardsAmount) : 0
+  const token = isDualRewards && doubleRewardToken.symbol !== 'ZERO' ? doubleRewardToken?.symbol : null
+  const parsedToken = token ? token : 'ZERO'
```
Comment (specialist0x) [inline src/components/earn/FarmBanner/index.tsx:51]:
```suggestion
  const parsedToken = token ?? 'ZERO'
```

File: src/components/earn/FarmBanner/index.tsx [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    [userFarmsAmount]
+  )
+
+  const totalUserTokensAmount = useMemo(() => {
+    if (countEnabled) {
+      const farmsArr = Object.values(userFarmsAmount)
+      const tokensArr = farmsArr.map(({ tokens }) => tokens)
+
+      const result = tokensArr.reduce((acc, n) => {
+        for (var prop in n) {
```
Comment (specialist0x) [inline src/components/earn/FarmBanner/index.tsx:157]: nit

```suggestion
        for (let prop in n) {
```

File: src/components/earn/FarmBanner/index.tsx [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                    <>
+                      <StyledSummaryHeader>Your total staked Amount: $</StyledSummaryHeader>
+                      {Number(TotalUserFarmsAmount).toFixed(2)}
+                    </>
+                  )}
+                </TYPE.subHeader>
+                {countEnabled && !isEmpty(totalUserTokensAmount) && (
+                  <>
+                    <TYPE.subHeader marginTop="0.7rem" marginBottom="0.7rem">
+                      <StyledSummaryHeader> Your total unreclaimed tokens:</StyledSummaryHeader>
```
Comment (specialist0x) [inline src/components/earn/FarmBanner/index.tsx:193]:
```suggestion
                      <StyledSummaryHeader> Your total unclaimed tokens:</StyledSummaryHeader>
```

Comment (specialist0x) [reply src/components/earn/FarmBanner/index.tsx:193]: Maybe "Claimable tokens:"

File: src/pages/EarnTri/index.tsx [hunk 5]
```diff
@@ -133,6 +133,8 @@ export default function Earn({
     return sortDescending ? <ChevronDown size={15} /> : <ChevronUp size={15} />
   }
 
+  const [userStakedAmount, setUsererStakedAmount] = useState(null)
+
```
Comment (specialist0x) [inline src/pages/EarnTri/index.tsx:137]: These look like they're unused

File: src/data/TotalStakedInPool.ts [hunk 6]
```diff
@@ -12,6 +12,7 @@ export function useTotalStakedInPool(token?: Token, version?: ChefVersions): Tok
     version === ChefVersions.V1 ? MASTERCHEF_ADDRESS_V1[ChainId.AURORA] : MASTERCHEF_ADDRESS_V2[ChainId.AURORA]
 
   const contract = useTokenContract(token?.address, false)
+
```
Comment (specialist0x) [inline src/data/TotalStakedInPool.ts:15]: This doesn't need to be part of this diff, right?

File: src/pages/EarnTri/Manage.tsx [hunk 7]
```diff
@@ -125,6 +125,7 @@ export default function Manage({
     totalStakedInUSD,
     doubleRewardToken
   } = stakingInfo
+
```
Comment (specialist0x) [inline src/pages/EarnTri/Manage.tsx:128]: This doesn't need to be part of this diff, right?

File: src/components/earn/FarmBanner/FarmBanner.types.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+export type FarmAmount = {
+  [id: number]: {
+    totalStaked: number
+    tokens: {
+      [id: string]: number
+    }
+  }
+}
```
Comment (specialist0x) [inline src/components/earn/FarmBanner/FarmBanner.types.ts:37]: Is there a reason this needs to be `key`'d? If these items were pushed as objects in to an array, could the portfolio balance show the sum of all items in the array

It'd make managing a flat data structure easier, too

File: src/components/earn/FarmBanner/index.tsx [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        setUserFarmsAmount(newUserFarmAmount)
+      }
+    }
+  }
+
+  useEffect(() => {
+    const userFarmsArr = Object.values(userFarmsAmount)
+    const allFarmsHaveInfo = !isEmpty(userFarmsArr) && userFarmsArr.every(farm => farm.totalStaked > 0)
+
+    if (!countEnabled) {
```
Comment (specialist0x) [inline src/components/earn/FarmBanner/index.tsx:124]: This seems hacky

Comment (specialist0x) [reply src/components/earn/FarmBanner/index.tsx:124]: If a pool exists its balance should be checked, no?

**Review by specialist0x — CHANGES_REQUESTED** [review]

**Comment by memacius32** [comment]
Replaced by #146, Closing.

**Pull request closed without merging**